### ISSUE TITLE
Dynamic Batching not creating batches correctly and incorrect inference results

### ISSUE THREAD

**Issue by omrifried** [triton-inference-server/server#4133]
**Description**
I am deploying a triton server to GKE via the [gke-marketplace-app](https://github.com/triton-inference-server/server/tree/main/deploy/gke-marketplace-app) documentation. When I try to use dynamic batching, the requests are not batched and it is only sent with a batch size of 1. Additionally, the inference only results in one detection when it should be multiple.

**Triton Information**
The version is 2.17 as this is what the marketplace feature deploys.

Are you using the Triton container or did you build it yourself?
Deployed via gcp marketplace

**To Reproduce**
I create the inference server with the following config:
```
name: "sample"
platform: "pytorch_libtorch"
max_batch_size : 16
input [
  {
    name: "INPUT__0"
    data_type: TYPE_UINT8
    format: FORMAT_NCHW
    dims: [ 3, 512, 512 ]
  }
]
output [
  {
    name: "OUTPUT__0"
    data_type: TYPE_FP32
    dims: [ -1, 4 ]
  },
 {
    name: "OUTPUT__1"
    data_type: TYPE_INT64
    dims: [ -1 ]
   label_filename: "sample.txt"
  },
  {
    name: "OUTPUT__2"
    data_type: TYPE_FP32
    dims: [ -1 ]
  }
]
 dynamic_batching {
    max_queue_delay_microseconds: 50000
}
```

I am calling inference as follows:
```
model = "sample"
client = httpclient.InferenceServerClient( url = url )

input_1 = httpclient.InferInput(name = "INPUT__0", shape = list(data.shape), datatype = "UINT8")
input_2 = httpclient.InferInput(name = "INPUT__0", shape = list(data.shape), datatype = "UINT8")

input_1.set_data_from_numpy(data, binary_data = True)
input_2.set_data_from_numpy(data, binary_data = True)

output_00 = httpclient.InferRequestedOutput(name = "OUTPUT__0", binary_data = False)
output_01 = httpclient.InferRequestedOutput(name = "OUTPUT__1", binary_data = False)
output_02 = httpclient.InferRequestedOutput(name = "OUTPUT__2", binary_data = False)

output_10 = httpclient.InferRequestedOutput(name = "OUTPUT__0", binary_data = False)
output_11 = httpclient.InferRequestedOutput(name = "OUTPUT__1", binary_data = False)
output_12 = httpclient.InferRequestedOutput(name = "OUTPUT__2", binary_data = False)

# Is this correct? I tried using reshape in the config, but it did not work. Without this I get errors about data shape.
input_1.set_shape([1, 3, 512, 512]
input_2.set_shape([1, 3, 512, 512]

response_1 = client.async_infer(model_name = model, inputs = [input_1], outputs = [output_00, output_01, output_02])
response_2 = client.async_infer(model_name = model, inputs = [input_2], outputs = [output_10, output_11, output_12])
```
  
**Expected behavior**
With the above code, when I run `print(response_1.get_result().get_response())` I am only seeing one detection, but I know that the model detects multiple objects during direct inference on local:
```
{... [{'name': 'OUTPUT__0', 'datatype': 'FP32', 'shape': [1, 4], 'data': [x_min, y_min, x_max, y_max]}, ...}
```
Additionally, when I run `print(client.get_inference_statistics())` I am seeing only a batch size of 1 when I expect 2 in this case:
```
{ ... 'batch_stats': [{'batch_size': 1, 'compute_input' : {'count': 2 ...}}] ... }
```

**Comment by omrifried** [comment]
I increased `max_queue_delay_microseconds `to `5000000`, and also tried removing it entirely, but I am still seeing `batch_size: 1`.

I set up my model as follows:
```
class TritonModel(torch.nn.Module):
   def __init__(self, model: object):
        super().__init__()
        self.model = model

   def transform(self, x: torch.Tensor):
        x = torch.squeeze(x)
        if x.dtype == torch.uint8:
             x /= 255
        return x

   def transform_output(self, prediction: Tuple[Dict[str, torch.Tensor], List[Dict[str, torch.Tensor]]]):
        # I believe this is what is causing the issue with the one detection return. All detections are returned without dynamic 
        # batching correctly, but when I add dynamic batching, I believe this is what is causing the issue.
        pred = prediction[1][0]
        out = tuple((pred['boxes'], pred['labels'], pred['scores']))
        return out

  def forward(self, x: torch.Tensor):
        x = self.transform(x)
        with torch.no_grad():
             out = self.transform_output(self.model[x])
             return out
```
I then compile with `triton_model = torch.jit.script(TritonModel(model).to(device)).to(device)` and upload this model along with the config mentioned above and a text file of classes. Then I call async_infer as mentioned above, but only see batch of 1.

Could this be due to the using version 2.17? Let me know if there is anything else needed from my end.

**Comment by tanmayv25** [comment]
Can you try a later releases? There have been a couple of fixes in Dynamic Batcher.
https://github.com/triton-inference-server/server/commit/7ad5141c93241c9bf385d11352d93a5cd380d738

**Comment by omrifried** [comment]
@tanmayv25  Yes, the marketplace tool uses 2.17 by default, but it looks like there is `build_and_push.sh` [here](https://github.com/triton-inference-server/server/tree/main/deploy/gke-marketplace-app/server-deployer). 

I want to confirm that the bash script is the correct way to push manually? Or is there a way to change the version on the marketplace tool?

**Comment by tanmayv25** [comment]
I am not familiar with that demo code. @mengdong Can you help?

**Comment by tanmayv25** [comment]
@omrifried  Meanwhile you can experiment with triton containers from NGC on your local machine to see if latest release resolves your issue.
Read more about the problem from here: https://github.com/triton-inference-server/server/issues/3624

**Comment by omrifried** [comment]
@tanmayv25 @mengdong it looks like the most up to date image available on GCP is the 2.17 version [here](https://console.cloud.google.com/gcr/images/nvidia-ngc-public/global/tritonserver)

**Comment by omrifried** [comment]
It looks like `build_and_push.sh` is the correct source to use to build the server from CLI. I see that [values.yaml](https://github.com/triton-inference-server/server/blob/main/deploy/gke-marketplace-app/server-deployer/chart/triton/values.yaml) reference gcr.io/nvidia-ngc-public/tritonserver.22.03-py3, but that image does not exist. Additionally, as mentioned above, I do not see where to label the existing cluster or namespace for the script. It would be great to deploy with an updated image using this bash script if possible.

**Comment by mengdong** [comment]
Hello @omrifried , you need to push the triton server from NCG to GCR, currently, GCR doesn't host Triton image.

**Comment by omrifried** [comment]
Understood. I tried this on my local with version 2.20 and am still seeing the same issue. I used the same code as I pasted above to call the server:
```
model = "sample"
client = httpclient.InferenceServerClient( url = url )

input_1 = httpclient.InferInput(name = "INPUT__0", shape = list(data.shape), datatype = "UINT8")
input_2 = httpclient.InferInput(name = "INPUT__0", shape = list(data.shape), datatype = "UINT8")

input_1.set_data_from_numpy(data, binary_data = True)
input_2.set_data_from_numpy(data, binary_data = True)

output_00 = httpclient.InferRequestedOutput(name = "OUTPUT__0", binary_data = False)
output_01 = httpclient.InferRequestedOutput(name = "OUTPUT__1", binary_data = False)
output_02 = httpclient.InferRequestedOutput(name = "OUTPUT__2", binary_data = False)

output_10 = httpclient.InferRequestedOutput(name = "OUTPUT__0", binary_data = False)
output_11 = httpclient.InferRequestedOutput(name = "OUTPUT__1", binary_data = False)
output_12 = httpclient.InferRequestedOutput(name = "OUTPUT__2", binary_data = False)

# Is this correct? I tried using reshape in the config, but it did not work. Without this I get errors about data shape.
input_1.set_shape([1, 3, 512, 512]
input_2.set_shape([1, 3, 512, 512]

response_1 = client.async_infer(model_name = model, inputs = [input_1], outputs = [output_00, output_01, output_02])
response_2 = client.async_infer(model_name = model, inputs = [input_2], outputs = [output_10, output_11, output_12])
```

**Comment by mengdong** [comment]
Hello @omrifried , sorry for the late response, before checking Triton config, I want to make sure your pytorch model supports dynamic batching. i.e. The model needs to have a batch dimension of -1. 

Triton dynamic batching feature requires the model to have -1 as batch size.

**Comment by tanmayv25** [comment]
Looks like a bug to me and worth investigating for a proper root cause. 
@omrifried Can you repeat your experiment with grpc client? Also, try sending 16 or more requests instead of just 2.

**Comment by omrifried** [comment]
I believe it is a cause of the way my model handles batching. Because it takes in a list of tensors, instead of a batch of tensors, dynamic batching will not work. I was able to get static batching to work by adjusting the forward method a bit, but I don't think I will be able to get dynamic batching to work due to the requirement for fixed length in torchscript. 

That being said, I did try with gRPC and larger sizes for dynamic batching (the same method I originally tried that was giving me only the first detection), and I was able to see the correct batch sizes when looking at inference statistics. So I think there still is a bug on the httpclient in terms of handling dynamic batching that the gRPC client does not have.

**Comment by kelisiya** [comment]
> I believe it is a cause of the way my model handles batching. Because it takes in a list of tensors dynamic batching will not work since it is an explicit list of tensors and not NCHW where N >= 1. I was able to get static batching to work by adjusting the forward method a bit, but I don't think I will be able to get dynamic batching to work due to the requirement for fixed length in torchscript when decomposing items.
> 
> That being said, I did try with gRPC and larger sizes for dynamic batching (the same method I originally tried that was giving me only the first detection), and I was able to see the correct batch sizes when looking at inference statistics. So I think there still is a bug on the httpclient in terms of handling dynamic batching that the gRPC client does not have.



> Looks like a bug to me and worth investigating for a proper root cause. @omrifried Can you repeat your experiment with grpc client? Also, try sending 16 or more requests instead of just 2.

I have the same problem. When I use torchscript model in my triton client , the  dynamic batching not work. How can I prove that the dynamic batching is being used correctly. Whether multiple requests correspond to one client or multiple requests correspond to multiple clients. I have an input image but it has to be sliced into 16 patches then inference these patches separately, I want triton to handle this case.

**Comment by tanmayv25** [comment]
> That being said, I did try with gRPC and larger sizes for dynamic batching (the same method I originally tried that was giving me only the first detection), and I was able to see the correct batch sizes when looking at inference statistics. So I think there still is a bug on the httpclient in terms of handling dynamic batching that the gRPC client does not have.

@omrifried  Yes. The httpclient is built on top of geventhttpclient lib. The greenlets needs co-operative multi-threading to actually relinquish control to one another. After:

```
response_1 = client.async_infer(model_name = model, inputs = [input_1], outputs = [output_00, output_01, output_02])
response_2 = client.async_infer(model_name = model, inputs = [input_2], outputs = [output_10, output_11, output_12])
```

Are you calling [get_results()](https://github.com/triton-inference-server/client/blob/main/src/python/library/tritonclient/http/__init__.py#L1559) on response_1 and response_2 to relinquish control? 

See here for example: https://github.com/triton-inference-server/client/blob/main/src/python/examples/simple_http_async_infer_client.py#L95

Because of this behavior our http client and grpc client APIs are different.

**Comment by tanmayv25** [comment]
> I have the same problem. When I use torchscript model in my triton client , the dynamic batching not work. How can I prove that the dynamic batching is being used correctly. Whether multiple requests correspond to one client or multiple requests correspond to multiple clients. I have an input image but it has to be sliced into 16 patches then inference these patches separately, I want triton to handle this case.

@kelisiya This looks like a separate problem. Can you create a new github issue?

**Comment by GuanLuo** [comment]
Also note that for HTTP Python client, you will need to set the `concurrency` for sending requests concurrently https://github.com/triton-inference-server/client/blob/main/src/python/examples/simple_http_async_infer_client.py#L55-L58

**Comment by omrifried** [comment]
@tanmayv25 - yes I am calling get_result sequentially, so it should be relinquishing control accordingly. 

I have also tried setting `concurrency` to a larger value, such as 10 or 20, but it does not seem to be helping there

**Comment by tanmayv25** [comment]
Interesting. I have opened a bug with our team and will try to look at it later.

**Comment by rmccorm4** [comment]
Hi @omrifried ,

Can you share a simple model that can reproduce the issue? It will help with looking into the root cause.

**Comment by omri-cavnue** [comment]
Hi @rmccorm4, I believe the issue is with the model architecture. For this test, I used the [Resnet50 FasterRCNN](https://pytorch.org/vision/stable/generated/torchvision.models.detection.fasterrcnn_resnet50_fpn.html) from PyTorch and converted it to a torchscript model.

This model takes in a list of tensors instead of a batch of tensors and I do not believe dynamic batching can be done due to this.

**Comment by omrifried** [comment]
Hi @rmccorm4, I believe the issue is with the model architecture. For this test, I used the [Resnet50 FasterRCNN](https://pytorch.org/vision/stable/generated/torchvision.models.detection.fasterrcnn_resnet50_fpn.html) from PyTorch and converted it to a torchscript model.

This model takes in a list of tensors instead of a batch of tensors and I do not believe dynamic batching can be done due to this.

**Comment by rmccorm4** [comment]
Hi @omrifried ,

Thanks for the reference. Do you mind sharing 

1. a script I can run as-is to generate the torchscript model (or share the model itself)? 
2. the corresponding Triton config.pbtxt to serve the model

This would help greatly to save time to look into this, thanks.

**Comment by dyastremsky** [comment]
Closing issue due to lack of activity. Please re-open the issue if you would like to follow up with this issue.
